torch/distributed/_mesh_layout.py

```diff
@@ -0,0 +1,337 @@
+        res_strides: list[int] = []
+        current_idx = 1
+        for size, stride in sorted(self.sizes_and_strides, key=lambda x: x[1]):
+            if stride == 0 or size == 1:
+                continue
+            assert current_idx <= size * stride, (
+                f"current_idx {current_idx} larger than numel so far {size * stride}."
+            )
+
+            res_sizes.append(stride // current_idx)
+            res_strides.append(current_idx)
+            current_idx = size * stride
+
+        res_sizes.append(self.ceil_div(world_size, current_idx))
+        res_strides.append(current_idx)
+        # This is different from original pycute implementation, because in pytorch we usually
+        # have the right-most dimension as the innermost dimension (smallest stride).
+        res_sizes.reverse()
+        res_strides.reverse()
+
+        return _Layout(tuple(res_sizes), tuple(res_strides)).coalesce()
+
+    def local_ranks(self) -> list[int]:
+        """
+        This function computes the local rank specified by the layout.
+
+        How it works:
+        1. we enumerates every possible coordinate (like a nested for-loop).
+        If sizes = (2, 3), we get the following coordinates:
+            (0,0), (0,1), (0,2), (1,0), (1,1), (1,2)
+
+        2. For each coordinate, we compute a linear rank index as:
+            local_ranks = sum(coord[i] * strides[i] for i in range(ndim))
+
+        Example A:
+        sizes   = (2, 3)        # 2 rows, 3 cols
+        strides = (3, 1)        # row-major layout
+        coords  = (0,0) -> 0*3 + 0*1 = 0
+                  (0,1) -> 0*3 + 1*1 = 1
+                  (0,2) -> 0*3 + 2*1 = 2
+                  (1,0) -> 1*3 + 0*1 = 3
+                  (1,1) -> 1*3 + 1*1 = 4
+                  (1,2) -> 1*3 + 2*1 = 5
+        result = [0, 1, 2, 3, 4, 5]
+
+        Example B:
+        sizes   = (2, 3)
+        strides = (1, 2)        # non-standard / strided layout
+        coords  = (0,0) -> 0*1 + 0*2 = 0
+                  (0,1) -> 0*1 + 1*2 = 2
+                  (0,2) -> 0*1 + 2*2 = 4
+                  (1,0) -> 1*1 + 0*2 = 1
+                  (1,1) -> 1*1 + 1*2 = 3
+                  (1,2) -> 1*1 + 2*2 = 5
+        result = [0, 2, 4, 1, 3, 5]
+        """
+        return [
+            sum(c * s for c, s in zip(coord, self.strides))
+            for coord in product(*(range(s) for s in self.sizes))
+        ]
+
+    def global_ranks(self, world_size: int) -> list[list[int]]:
+        """
+        Build global ranks specified by the layout via two-level ranks composition.
+
+        The nested list forms the Cartesian product of group ranks and group offset
+        and the final global ranks are the addition of these two. The result is a
+        list of lists: one sublist per group. This rank list will be used to build
+        the communicator underlying the layout.
+
+        Example:
+        group ranks    = [0, 1, 2, 3]
+        group offsets  = [0, 4, 8, 12]
```

Comment:
For the example here, it would be much clearer if it was also specified what world_size and self.size/stride were, that generated this case.